File: C#/myLeetCode/Array/59. Spiral Matrix II.cs
// 59. Spiral Matrix II
// https://leetcode.com/problems/spiral-matrix-ii/

using System;

public class Solution {
    public int[][] GenerateMatrix(int n) {
        int[][] matrix = new int[n][];
        for (int i = 0; i < n; i++) {
            matrix[i] = new int[n];
        }
        
        int left = 0, top = 0;
        int right = n - 1, bottom = n - 1;
        int num = 1;

        while (left <= right && top <= bottom) {
            for (int i = left; i <= right; i++) {
                matrix[top][i] = num++;
            }
            for (int i = top + 1; i <= bottom; i++) {
                matrix[i][right] = num++;
            }
            if (left < right && top < bottom) {
                for (int i = right - 1; i > left; i--) {
                    matrix[bottom][i] = num++;
                }
                for (int i = bottom; i > top; i--) {
                    matrix[i][left] = num++;
                }
            }
            left++;
            right--;
            top++;
            bottom--;
        }
        return matrix;
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();
        int n = 3; // You can change this to test other sizes.
        int[][] result = solution.GenerateMatrix(n);

        // Print the matrix to the console
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                Console.Write(result[i][j] + "\t");
            }
            Console.WriteLine();
        }
    }
}

File: C++/myLeetCode/Array/59. Spiral Matrix II.cpp
// 59. Spiral Matrix II
// https://leetcode.com/problems/spiral-matrix-ii/

#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> generateMatrix(int n) {
        std::vector<std::vector<int>> matrix(n, std::vector<int>(n));
        int num = 1;
        int row = 0, col = 0;
        std::vector<std::vector<int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int dir = 0;

        while (num <= n * n) {
            if (row >= 0 && row < n && col >= 0 && col < n && matrix[row][col] == 0) {
                matrix[row][col] = num++;
                row += directions[dir][0];
                col += directions[dir][1];
            } else {
                // 回退一步并改变方向
                row -= directions[dir][0];
                col -= directions[dir][1];
                dir = (dir + 1) % 4;
                row += directions[dir][0];
                col += directions[dir][1];
            }
        }
        return matrix;
    }
};

int main() {
    Solution sol;
    int n = 4;
    std::vector<std::vector<int>> result = sol.generateMatrix(n);

    // 打印结果
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            std::cout << result[i][j] << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}
